core/store/src/flat/storage.rs

```diff
@@ -177,8 +181,11 @@ impl FlatStorageInner {
             if new_head == current_flat_head_hash {
                 return Ok(current_flat_head_hash);
             }
-            let metadata =
-                self.deltas.get(&new_head).ok_or(missing_delta_error(&new_head))?.metadata;
+            let metadata = self
+                .deltas
+                .get(&new_head)
+                .ok_or(self.create_block_not_supported_error(&new_head))?
```

Comment:
Can you document it in the method comment? Or anywhere else more appropriate. Treating error types as control flow mechanism is not that typical pattern so it's worth pointing it out. An alternative would be to move that information from the error to the return type instead.